# Assembly evaluation pipeline

A pipeline used to evaluate _de novo_ genome assemblies. Theoretically can be used on both contigs and scaffolds. It aggregates reports from several independent evaluation methods. It is tailored towards the evaluation of assemblies generated using solely PacBio HiFi data. It will also attempt to extract mitochondrial genome, decontaminate, and correct misassembly events.

![](assembly_qc/dag.pdf)

### Metrics evaluated:
* Genome gene complement completeness by utilizing [BUSCO](https://busco.ezlab.org/).
* Genome k-mer completeness by utilizing [merqury](https://github.com/marbl/merqury).
* Genome correctness based on read evidence by utilizing [inspector](https://github.com/Maggi-Chen/Inspector). Includes a custom configuration of [inspector](https://github.com/Maggi-Chen/Inspector) to handle non-standard FASTA headers (e.g., the `|` symbol added by `hifiasm`).
* Genome contamination levels by utilizing NCBI's Foreign Contamination Screening ([fcs](https://github.com/ncbi/fcs)).

### Genome cleaning steps:
* It uses [inspector](https://github.com/Maggi-Chen/Inspector) correction module. It uses read alignments to correct smaller misassembly events.
* Using the contaminations identified through [fcs](https://github.com/ncbi/fcs) it will remove contigs flagged as contaminants. It uses the `EXCLUDE` tag generated by [fcs](https://github.com/ncbi/fcs).
* Using [MitoHIFI](https://github.com/marcelauliano/MitoHiFi) it will detect mitochondrial genome in the assembly, and will separate it.

> **NOTE**  
> 
> For the different tools I used the parameters I deemed appropriate for my own analyses. To customize the pipeline towards specific needs, please inqure the documentation of the desired tool and adjust the `Snakefile` accordingly.
## Usage

### Prerequisites
The pipeline requires [Snakemake](https://snakemake.readthedocs.io/en/stable/) and [Anaconda](https://docs.anaconda.com/miniconda/).

### Databases
Several databases need to be downloaded:
The FCS database can be attained by following the [fcs-gx](https://github.com/ncbi/fcs/wiki/FCS-GX-input#fcs-gx-database-location) guide (see below).
```
SOURCE_DB_MANIFEST="https://ftp.ncbi.nlm.nih.gov/genomes/TOOLS/FCS/database/test-only/test-only.manifest"
LOCAL_DB="/path/to/db/folder"
python3 fcs.py db get --mft "$SOURCE_DB_MANIFEST" --dir "$LOCAL_DB/test-only" 
```

The BUSCO lineage should be *a priori* downloaded and be present in the working directory of the Snakefile with `busco_downloads/lineages/[user_determined_lineage]`.

> **CAUTION**
> 
> The pipeline will only choose the first directory if multiple BUSCO lineage directories exist within the path.

One can access lineages with the following commands.
```
# To view available lineages
busco --list-datasets

# To download the desired lineage
busco --download [user_determined_lineage]
```

### Data location
The data should be located in:
* Assemblies with a `.fa` suffix should be placed in `workflow/resources/assemblies/` directory.
* Concatenated PacBio HiFi reads in a **fasta** format should be placed in `workflow/resources/reads/` directory. 

> **NOTE**
> 
> The pipeline is intended to evaluate different _de novo_ genome assemblies originating from the same reads. It **cannot** evaluate assemblies from different species. For this one needs to setup different working directories separately for each species.

### Config setup
In the `config/config.yml` the user needs to set the following parameters:
* Which reads to utilize (eg. `workflow/resources/reads/reads.fa.gz`).
* The `gxdb` path used by [fcs-gx](https://github.com/ncbi/fcs/wiki/FCS-GX-input#fcs-gx-database-location).
* The taxonomic ID of the species for which the different assemblies were generated (use the [NCBI Taxonomy Database](https://www.ncbi.nlm.nih.gov/taxonomy))
* Species name should also be included in quotation marks.
* An estimated genome size is required. Preferably the estimate is backed up by empirical data. If estimates based solely on k-mer spectra are available, then the user should take extra care to evaluate the model fit.
* An estimated coverage of the reads is also expected. This can be easily calculated with `total bases in reads / genome size`. 

### Usage

With the prerequisites done, simply run:

```
snakemake --cores [user-defined] --use-conda all
```

## Output

The runs from different evaluation methods are grouped in `[method]_report/` directories. Within each directory multiple directories are found for each input assembly.

For each input assembly an output directory is created in `final_assemblies/`. Within each output directory users will find the output from [MitoHIFI](https://github.com/marcelauliano/MitoHiFi), and the de-contaminated, corrected `contig_corrected.fa` final assembly. 

> **NOTE**
> 
> To facilitate comparison between assemblies, a summary script collects the outputs from all evaluation methods. The aggregated results are saved in `assembly_evaluation/assembly_metrics_summary.csv`, along with quick-look plots.



